src/DistIL/IR/MatchExtensions.cs

```diff
@@ -0,0 +1,254 @@
+    {
+        bool result = true;
+        if (typed.Argument is not null) {
+            result = MatchArgument(value, typed.Argument, outputs);
+        }
+
+        if (typed.Type is "const") {
+            result &= value is Const;
+        } else if (typed.Type is "instr") {
+            result &= value is Instruction;
+        } else {
+            result &= PrimType.GetFromAlias(typed.Type) == value.ResultType;
+        }
+
+        return result;
+    }
+
+    private static bool MatchValue(Value value, IInstructionPatternArgument pattern, OutputPattern outputs)
+    {
+        return pattern switch {
+            InstructionPattern p when value is Instruction instruction => MatchInstruction(instruction, p, outputs),
+            _ => MatchArgument(value, pattern, outputs)
+        };
+    }
+
+    private static bool MatchConstArgument(ConstantArgument constantArg, Const constant)
+    {
+        if (constantArg.Type == constant.ResultType) {
+            if (constantArg is StringArgument strArg) {
+                return MatchStringArg(strArg, constant as ConstString);
+            }
+
+            object? value = constant switch {
+                ConstInt constInt => constInt.Value,
+                ConstFloat constFloat => constFloat.Value,
+                ConstNull => null,
+                _ => null
+            };
+
+            return value.Equals(constantArg.Value);
+        }
+
+        return false;
+    }
+
+    private static bool MatchStringArg(StringArgument strArg, ConstString constant)
+    {
+        if (strArg.Operation == StringOperation.StartsWith) {
+            return constant.Value.StartsWith(strArg.Value.ToString()!);
+        }
+        if (strArg.Operation == StringOperation.EndsWith) {
+            return constant.Value.EndsWith(strArg.Value.ToString()!);
+        }
+        if (strArg.Operation == StringOperation.Contains) {
+            return constant.Value.Contains(strArg.Value.ToString()!);
+        }
+
+        return strArg.Value.Equals(constant.Value);
+    }
+
+
+    private static bool MatchBinary(BinaryInst bin, InstructionPattern pattern, OutputPattern outputs)
+    {
+        var operation = pattern.OpCode;
+        var op = (BinaryOp)(operation - (Opcode._Bin_First + 1));
+
+        if (bin.Op != op) {
+            return false;
+        }
+
+        return MatchOperands(bin, pattern, outputs);
+    }
+
+    private static bool MatchCompare(CompareInst comp, InstructionPattern pattern, OutputPattern outputs)
+    {
+        var operation = pattern.OpCode;
+        var op = (CompareOp)(operation - (Opcode._Cmp_First + 1));
+
+        if (comp.Op != op) {
+            return false;
+        }
+
+        return MatchOperands(comp, pattern, outputs);
+    }
+
+    private static bool MatchUnary(UnaryInst un, InstructionPattern pattern, OutputPattern outputs)
+    {
+        var operation = pattern.OpCode;
+        var op = (UnaryOp)(operation - (Opcode._Bin_First + 1));
```

Comment:
We need to add the unary opcodes in the enum, otherwise this should probably be commented as todo.

(also this enum is a mess, purely my fault, it would probably be nice to have helpers like `Get/Is BinaryOp()` and etc.)